cineast-api/src/main/java/org/vitrivr/cineast/api/websocket/handlers/queries/TemporalQueryMessageHandlerV2.java

```diff
@@ -0,0 +1,253 @@
+              segments.addAll(stageSegments);
+              stageResults.addAll(results);
+
+              /* We limit the results to be sent back to the requester to the max limit. This is so that the original view is not affected by the changes of temporal query version 2 */
+              List<StringDoublePair> limitedResults = results.stream()
+                  .limit(max)
+                  .collect(Collectors.toList());
+              results.forEach(res -> limitedRelevantSegments.add(res.key));
+              List<String> limitedSegmentIds = limitedResults.stream()
+                  .map(el -> el.key)
+                  .collect(Collectors.toList());
+              sentSegmentIds.addAll(limitedSegmentIds);
+              List<String> limitedObjectIds = this.submitSegmentAndObjectInformation(session, uuid, limitedSegmentIds);
+              sentObjectIds.addAll(limitedObjectIds);
+              futures.addAll(this.finalizeAndSubmitResults(session, uuid, category, qc.getContainerId(), limitedResults));
+              List<Thread> _threads = this.submitMetadata(session, uuid, limitedSegmentIds, limitedObjectIds, segmentIdsForWhichMetadataIsFetched, objectIdsForWhichMetadataIsFetched);
+              metadataRetrievalThreads.addAll(_threads);
+            }
+          }
+        }
+
+        /* After having finished a stage, we add all relevant segments to the config of the next stage. */
+        if (relevantSegments.size() == 0) {
+          LOGGER.warn("No relevant segments anymore, aborting staged querying");
+          /* Clear the relevant segments are there are none */
+          stageQConf.setRelevantSegmentIds(relevantSegments);
+          break;
+        }
+        stageQConf.setRelevantSegmentIds(relevantSegments);
+        relevantSegments.clear();
+      }
+      limitedStageQConf.setRelevantSegmentIds(limitedRelevantSegments);
+
+      /* At this point, we have iterated over all stages. Now, we need to go back for all stages and send the results for the relevant ids. */
+      for (int stageIndex = 0; stageIndex < stagedSimilarityQuery.getStages().size() - 1; stageIndex++) {
+        int finalContainerIdx = containerIdx;
+        int finalStageIndex = stageIndex;
+        /* Add the results from the last filter from all previous stages also to the list of results */
+        cache.get(stageIndex).forEach((category, results) -> {
+          results.removeIf(pair -> !stageQConf.getRelevantSegmentIds().contains(pair.key));
+          stageResults.addAll(results);
+        });
+        /* Return the limited results from all stages that are within the filter */
+        cache.get(stageIndex).forEach((category, results) -> {
+          results.removeIf(pair -> !limitedStageQConf.getRelevantSegmentIds().contains(pair.key));
+          Thread thread = new Thread(() -> {
+            futures.addAll(this.finalizeAndSubmitResults(session, uuid, category, finalContainerIdx, results));
```

Comment:
I don't understand the reasoning for using threads here. Doesn't this cause concurrency issues since multiple threads are using the web socket at the same time? Is there really a benefit from introducing asynchronicity here?

Reply:
The reasoning is that a lot of time is used serializing to JSON, and putting this in a thread enables us to move forward with the main logic. I don’t have numbers at hand, but in my benchmarking last summer this change sped up walltime significantly.

Reply:
Interesting. In that case, we might want to think about using a dedicated thread pool for these sorts of things in order to be able to configure the thread volume used.

Reply:
Is there anything I can do about this today or can I set it as resolved for the purpose of this pull request?

Reply:
No, this is more of a general comment, something to keep in mind for future revisions.